pandas/core/arrays/categorical.py

```diff
@@ -400,20 +400,39 @@ def astype(self, dtype: Dtype, copy: bool = True) -> ArrayLike:
             If copy is set to False and dtype is categorical, the original
             object is returned.
         """
-        if is_categorical_dtype(dtype):
+        if self.dtype is dtype:
+            result = self.copy() if copy else self
+
+        elif is_categorical_dtype(dtype):
             dtype = cast(Union[str, CategoricalDtype], dtype)
 
             # GH 10696/18593
             dtype = self.dtype.update_dtype(dtype)
             self = self.copy() if copy else self
-            if dtype == self.dtype:
-                return self
-            return self._set_dtype(dtype)
-        if is_extension_array_dtype(dtype):
-            return array(self, dtype=dtype, copy=copy)
-        if is_integer_dtype(dtype) and self.isna().any():
+            result = self._set_dtype(dtype)
+
+        elif is_extension_array_dtype(dtype):
+            result = array(self, dtype=dtype, copy=copy)
+
+        elif is_integer_dtype(dtype) and self.isna().any():
             raise ValueError("Cannot convert float NaN to integer")
-        return np.array(self, dtype=dtype, copy=copy)
+
+        elif len(self.codes) == 0 or len(self.categories) == 0:
+            result = np.array(self, dtype=dtype, copy=copy)
+
+        else:
+            # GH8628 (PERF): astype category codes instead of astyping array
+            try:
+                astyped_categories = np.array(
+                    self.categories.to_numpy(), dtype=dtype, copy=copy
```

Comment:
i think do `self.categories.astype` and then cast to numpy if necessary afterwards, may avoid an object-dtype cast in the middle

Reply:
Yeah that cuts the time. Before:
```
In [2]: %timeit [df[col].astype('unicode') for col in df.columns]
106 ms ± 28.1 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
and after
```
In [2]: %timeit [df[col].astype('unicode') for col in df.columns]
52.8 ms ± 4.77 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```